### PR TITLE
[POC] Validate extension update manifest signed with a certificated public key

### PR DESCRIPTION
**Pull request by alikon**
Pull Request for Issue #19792 . (it cover only the extension update manifest signature/verification)

### Summary of Changes
suppose we can trust joomla as a certification autority
so suppose joomla is able to produce a keypair (ca secret key, ca public key) in a very secure way and deliver the ca public key in the joomla code

an extension developer is able to produce his keypairs (developer secret key, developer public key)
the extension developer ask the joomla CA to sign his developer public key (digital certificate)
the joomla ca hash the developer public key and sign the digest with her secret key and give back to the developer the hashed/signed developer public key

the extension developer embed in his update extension manifest the  developer public key , and the  joomla ca signed developer public key

the extension developer sign the hash-384 of his update extension manifest with his developer private key lets call this signed digest manifest

the extension developer embed in his update extension manifest the signed digest manifest , 

the update verification process

get the developer public key and the signed digest manifest  and the  joomla ca signed developer public key from the extension update manifest

check if we can trust the developer public key against the joomla ca public key from the joomla code

remove `<signature>` from the extension update manifest

calculate the sha-384 of the extension update manifest

verify the manifest digest against the developer public key

### Testing Instructions
0. Generate a keypairs for the JOOMLA! Certificate Autority (with this gist https://gist.github.com/alikon/db3356e8bdcc3c23968cec1179870302)
1. Generate a keypairs for the extension developer (with the same gist at point 0.)

2. Change the JOOMLA! Certificate Autority public key to the /site/administrator/components/com_installer/jcapk.xml with the one generated at step 0.

3. Ask to the JOOMLA! Certificate Autority to certificate your extension developer public key generated at step 1. (with this gist https://gist.github.com/alikon/f7e22c5f474637f4c1c815d9c5481a72)

4. Add to your extension update manifest your developer public key and your certificate

5. Generate your signed extension update manifest (with this gist https://gist.github.com/alikon/233d1111cdd2e0d0b1cec18596bbc152)

6. Publish it to your update server
should be something like this (i've used the com_patchtester manifest)
```xml
<?xml version="1.0" encoding="utf-8"?>
<updates>
	<update>
		<name>Patch Tester Component</name>
		<description>Joomla! CMS Patch Tester Component</description>
		<element>com_patchtester</element>
		<type>component</type>
		<version>2.0.1</version>
		<client>administrator</client>
		<infourl title="Patch Tester Component">https://github.com/joomla-extensions/patchtester/releases/tag/2.0.1</infourl>
		<downloads>
			<downloadurl type="full" format="zip">https://github.com/joomla-extensions/patchtester/releases/download/2.0.1/com_patchtester.zip</downloadurl>
		</downloads>
		<tags>
			<tag>stable</tag>
		</tags>
		<targetplatform name="joomla" version="3.[56789]" />
	</update>
	<update>
		<name>Patch Tester Component</name>
		<description>Joomla! CMS Patch Tester Component</description>
		<element>com_patchtester</element>
		<type>component</type>
		<version>3.0.0-beta4</version>
		<client>administrator</client>
		<infourl title="Patch Tester Component">http://localhost/test/</infourl>
		<downloads>
			<downloadurl type="full" format="zip">http://localhost/test/com_patchtester.tar.bz2</downloadurl>
		</downloads>

		<tags>
			<tag>beta</tag>
		</tags>
		<targetplatform name="joomla" version="3.[789]" />
	</update>
	<publickey>6c46e9505ba78f9fb113b34dace5c095f69d1f693d5a087272cf6d470e5e15c9</publickey>
	<certificate>06cf537b63314038f1a18a928805b40aa5fbe3d01eaf849eb0c5d624be0ef3587909dfb0e1fce12a66ff8c2bbfa769d516f1f1932aae3f515e5fe0f94dc39b00</certificate>
	<signature>b003efd584e43d66daeaddd754ccd4c21f2a34270dccfcabcfa3d56f48c3eb822afd29a0981df2c42ff6ac0cf6dcafc1605ba2b9d1bce033f4245d32d137fd00</signature>
</updates>
```
7. In Extension->Manage->Update Select the extension update items and check-it clicking on the Verify button

### Expected result
the extension update manifest is validated against developer public key wich is certificated by the Joomla CA

### Actual result
no extension update manifest verification

### Documentation Changes Required

the Joomla CA public key is showed on System->System Information
screenshot from 2018-04-03 19-16-16

the extension update manifest verification
Extensions->Manage->Update
screenshot from 2018-04-03 19-20-34

### CODE REVIEW

File: administrator/components/com_installer/models/update.php [hunk 1]
```diff
@@ -616,4 +616,132 @@ protected function preparePreUpdate($update, $table)
...
+	
+			// Remove the signature
+			$doc = new DOMDocument;
+			$doc->preserveWhiteSpace = false;
+			$dom->formatOutput = true;
+			$doc->loadxml( $file );
+			$xpath = new DOMXPath($doc);
+			foreach( $xpath->query("//signature") as $node) 
+			{
+    			$node->parentNode->removeChild($node);
```
Comment (Quy) [inline administrator/components/com_installer/models/update.php]: Indent

File: administrator/components/com_installer/models/update.php [hunk 2]
```diff
@@ -616,4 +616,132 @@ protected function preparePreUpdate($update, $table)
...
+
+			$file = file_get_contents($instance->detailsurl);
+	
+			// Remove the signature
+			$doc = new DOMDocument;
+			$doc->preserveWhiteSpace = false;
+			$dom->formatOutput = true;
+			$doc->loadxml( $file );
+			$xpath = new DOMXPath($doc);
+			foreach( $xpath->query("//signature") as $node)
```
Comment (Quy) [inline administrator/components/com_installer/models/update.php]: Add blank line above and add space after `foreach` and remove space after `(`.

File: administrator/components/com_installer/models/update.php [hunk 3]
```diff
@@ -616,4 +616,132 @@ protected function preparePreUpdate($update, $table)
...
+			$dom->formatOutput = true;
+			$doc->loadxml( $file );
+			$xpath = new DOMXPath($doc);
+			foreach( $xpath->query("//signature") as $node) 
+			{
+    			$node->parentNode->removeChild($node);
+			}
+			// Write the unsigned manfifest
+			$xml = $doc->savexml();
+ 			$a= file_put_contents(dirname(__DIR__) . '/unsignedmanifest.xml', $xml);
```
Comment (Quy) [inline administrator/components/com_installer/models/update.php]: Add space before `=`

File: administrator/components/com_installer/models/update.php [hunk 4]
```diff
@@ -616,4 +616,132 @@ protected function preparePreUpdate($update, $table)
...
+
+			// Read the unsigned manifest
+			$manifest = file_get_contents(dirname(__DIR__) . '/unsignedmanifest.xml'); 
+			// Calculate hash digest
+			$digest   = hash("sha384", $manifest);
+	
+			// Check the signature
+			if (ParagonIE_Sodium_Compat::crypto_sign_verify_detached($signature, $digest, $publickey)) {
```
Comment (Quy) [inline administrator/components/com_installer/models/update.php]: Move `{` to its own line.

File: administrator/components/com_installer/models/update.php [hunk 5]
```diff
@@ -616,4 +616,132 @@ protected function preparePreUpdate($update, $table)
...
+
+			// Read the unsigned manifest
+			$manifest = file_get_contents(dirname(__DIR__) . '/unsignedmanifest.xml'); 
+			// Calculate hash digest
+			$digest   = hash("sha384", $manifest);
+	
+			// Check the signature
+			if (ParagonIE_Sodium_Compat::crypto_sign_verify_detached($signature, $digest, $publickey)) { 
+				$app->enqueueMessage(\JText::sprintf('COM_INSTALLER_MANIFEST_VALID', $instance->name), 'message');
+			} else {
```
Comment (Quy) [inline administrator/components/com_installer/models/update.php]: Move `else` and `{` to its own line.

File: administrator/components/com_admin/models/sysinfo.php [hunk 6]
```diff
@@ -318,8 +319,28 @@ public function &getInfo()
 			'version'               => $version->getLongVersion(),
 			'platform'              => $platform->getLongVersion(),
 			'useragent'             => isset($_SERVER['HTTP_USER_AGENT']) ? $_SERVER['HTTP_USER_AGENT'] : '',
+			'jcapk4dev'             => $jcapk4dev,
 		);
 
+		// Get the Joomla CA public key for developers
+		$file = file_get_contents(JPATH_BASE . '/components/com_installer/jcapk.xml');
+		// Check the file
```
Comment (Quy) [inline administrator/components/com_admin/models/sysinfo.php:328]: Add blank line above.

File: administrator/components/com_admin/models/sysinfo.php [hunk 7]
```diff
@@ -318,8 +319,28 @@ public function &getInfo()
...
+		if (!$file)
+		{
+			return $this->info;
+		}
+		$xml  = new SimpleXMLElement($file);
```
Comment (Quy) [inline administrator/components/com_admin/models/sysinfo.php:334]: Add blank line above.

File: administrator/components/com_admin/models/sysinfo.php [hunk 8]
```diff
@@ -318,8 +319,28 @@ public function &getInfo()
...
+		// Parse the xml file.
```
Comment (Quy) [inline administrator/components/com_admin/models/sysinfo.php:336]: Add blank line above.

File: administrator/components/com_admin/models/sysinfo.php [hunk 9]
```diff
@@ -318,8 +319,28 @@ public function &getInfo()
...
+		if (!$xml)
+		{
+			return $this->info;
+		}
+		// Check the tag
```
Comment (Quy) [inline administrator/components/com_admin/models/sysinfo.php:342]: Add blank line above.

File: administrator/components/com_admin/models/sysinfo.php [hunk 10]
```diff
@@ -318,8 +319,28 @@ public function &getInfo()
...
+		if (!isset($xml->pk4dev))
+		{
+			return $this->info;
+		}
+		$this->info["jcapk4dev"] = (string) $xml->pk4dev;
```
Comment (Quy) [inline administrator/components/com_admin/models/sysinfo.php:348]: Add blank line above.

File: administrator/components/com_admin/models/sysinfo.php [hunk 11]
```diff
@@ -318,8 +319,28 @@ public function &getInfo()
...
 		return $this->info;
```
Comment (Quy) [inline administrator/components/com_admin/models/sysinfo.php:349]: Add blank line above.

File: administrator/components/com_installer/models/update.php [hunk 12]
```diff
@@ -616,4 +616,132 @@ protected function preparePreUpdate($update, $table)
...
+		if (!isset($xml->pk4dev))
+		{
+			return $isvalid;
+		}
+
+		$joomlaPK = ParagonIE_Sodium_Compat::hex2bin((string) $xml->pk4dev);
+		// Hash the developer public key
+		$digest      = ParagonIE_Sodium_Compat::hex2bin(hash("sha384", $developerPK));
+		$certificate = ParagonIE_Sodium_Compat::hex2bin($certificate);
+		// Check the certificate
```
Comment (Quy) [inline administrator/components/com_installer/models/update.php]: Add blank line above.

File: administrator/components/com_installer/models/update.php [hunk 13]
```diff
@@ -616,4 +616,132 @@ protected function preparePreUpdate($update, $table)
...
+		{
+			return $isvalid;
+		}
+		$xml  = new SimpleXMLElement($file);
+		// Parse the xml file.
+		if (!$xml)
+		{
+			return $isvalid;
+		}
+		// Check the tag
```
Comment (Quy) [inline administrator/components/com_installer/models/update.php]: Add blank line above.

File: administrator/components/com_installer/models/update.php [hunk 14]
```diff
@@ -616,4 +616,132 @@ protected function preparePreUpdate($update, $table)
...
+	{
+		$isvalid = false;
+		// Get the Joomla CA public key for developers
+		$file = file_get_contents(dirname(__DIR__) . '/jcapk.xml');
+		if (!$file)
+		{
+			return $isvalid;
+		}
+		$xml  = new SimpleXMLElement($file);
+		// Parse the xml file.
```
Comment (Quy) [inline administrator/components/com_installer/models/update.php]: Add blank line above.

File: administrator/components/com_installer/models/update.php [hunk 15]
```diff
@@ -616,4 +616,132 @@ protected function preparePreUpdate($update, $table)
...
+	protected function validateCertificate($certificate, $developerPK)
+	{
+		$isvalid = false;
+		// Get the Joomla CA public key for developers
+		$file = file_get_contents(dirname(__DIR__) . '/jcapk.xml');
+		if (!$file)
+		{
+			return $isvalid;
+		}
+		$xml  = new SimpleXMLElement($file);
```
Comment (Quy) [inline administrator/components/com_installer/models/update.php]: Add blank line above.

File: administrator/components/com_installer/models/update.php [hunk 16]
```diff
@@ -616,4 +616,132 @@ protected function preparePreUpdate($update, $table)
...
+	 * @return  boolean
+	 *
+	 * @since   __DEPLOY_VERSION__
+	 */
+	protected function validateCertificate($certificate, $developerPK)
+	{
+		$isvalid = false;
+		// Get the Joomla CA public key for developers
+		$file = file_get_contents(dirname(__DIR__) . '/jcapk.xml');
+		if (!$file)
```
Comment (Quy) [inline administrator/components/com_installer/models/update.php]: Add blank line above.

File: administrator/language/en-GB/en-GB.com_installer.ini [hunk 17]
```diff
@@ -86,6 +86,12 @@ COM_INSTALLER_LANGUAGES_FILTER_SEARCH_DESC="Search in language name and language
 COM_INSTALLER_LANGUAGES_FILTER_SEARCH_LABEL="Search Languages"
 COM_INSTALLER_MANAGE_FILTER_SEARCH_DESC="Search in extension name. Prefix with ID: to search for an extension ID."
 COM_INSTALLER_MANAGE_FILTER_SEARCH_LABEL="Search Extensions"
+COM_INSTALLER_MANIFEST_INVALID="The %s update manifest is not valid"
+COM_INSTALLER_MANIFEST_INVALID_CERTIFICATE="The %s update manifest have an untrusted Certificate"
```
Comment (Quy) [inline administrator/language/en-GB/en-GB.com_installer.ini]: Change all `have` to `has`.

**Comment by alikon** [comment]
@Quy i will do all the needed cs changes after and if we can get some consensus on the main concept
consider this for now as "workable playground"

**Comment by zero-24** [comment]
@alikon please take a look into the drone CS errors: http://ci.joomla.org/joomla/joomla-cms/4352/4 ;)

**Comment by mbabker** [comment]
This is a proof of concept, not a production pull request.  It doesn't really matter at this moment whether CI is passing.

File: administrator/components/com_installer/models/update.php [hunk 18]
```diff
@@ -616,4 +616,132 @@ protected function preparePreUpdate($update, $table)
...
+
+			$publickey = ParagonIE_Sodium_Compat::hex2bin($update->publickey->_data);
+			$signature = ParagonIE_Sodium_Compat::hex2bin($update->signature->_data);
+
+			$file = file_get_contents($instance->detailsurl);
+	
+			// Remove the signature
+			$doc = new DOMDocument;
+			$doc->preserveWhiteSpace = false;
+			$dom->formatOutput = true;
```
Comment (Quy) [inline administrator/components/com_installer/models/update.php]: `$doc`?

File: administrator/language/en-GB/en-GB.com_admin.ini [hunk 19]
```diff
@@ -22,6 +22,7 @@ COM_ADMIN_EXTENSIONS="Extensions"
 COM_ADMIN_FILE_UPLOADS="File Uploads"
 COM_ADMIN_GLOSSARY="Glossary"
 COM_ADMIN_GO="Go"
+COM_ADMIN_JCA_PUBLICKEY="Joomla CA PublicKey"
```
Comment (brianteeman) [inline administrator/language/en-GB/en-GB.com_admin.ini]: please change this to 
+COM_ADMIN_JCA_PUBLICKEY="Joomla CA Public Key"

Comment (alikon) [reply administrator/language/en-GB/en-GB.com_admin.ini]: done

File: administrator/language/en-GB/en-GB.com_installer.ini (hunk 17 above)
Comment (alikon) [reply administrator/language/en-GB/en-GB.com_installer.ini]: done

File: administrator/components/com_admin/models/sysinfo.php (hunk 11 above)
Comment (alikon) [reply administrator/components/com_admin/models/sysinfo.php:349]: done

File: administrator/components/com_admin/models/sysinfo.php (hunk 10 above)
Comment (alikon) [reply administrator/components/com_admin/models/sysinfo.php:348]: done

File: administrator/components/com_admin/models/sysinfo.php (hunk 9 above)
Comment (alikon) [reply administrator/components/com_admin/models/sysinfo.php:342]: done

File: administrator/components/com_admin/models/sysinfo.php (hunk 8 above)
Comment (alikon) [reply administrator/components/com_admin/models/sysinfo.php:336]: done

File: administrator/components/com_admin/models/sysinfo.php (hunk 7 above)
Comment (alikon) [reply administrator/components/com_admin/models/sysinfo.php:334]: done

File: administrator/components/com_admin/models/sysinfo.php (hunk 6 above)
Comment (alikon) [reply administrator/components/com_admin/models/sysinfo.php:328]: done

File: administrator/components/com_installer/controllers/update.php [hunk 20]
```diff
@@ -197,4 +197,26 @@ public function ajax()
...
+	public function check()
+	{
+		// Check for request forgeries.
+		JSession::checkToken() or jexit(JText::_('JINVALID_TOKEN'));
+
+		/** @var InstallerModelUpdate $model */
+		$model = $this->getModel('update');
+		$uid   = $this->input->get('cid', array(), 'array');
+		$uid   = ArrayHelper::toInteger($uid, array());
+
```
Comment (Quy) [inline administrator/components/com_installer/controllers/update.php:217]: Remove tab

File: administrator/components/com_installer/models/update.php [hunk 21]
```diff
@@ -616,4 +616,144 @@ protected function preparePreUpdate($update, $table)
...
+		$result = true;
+		$app = JFactory::getApplication();
+
+		foreach ($uids as $uid)
+		{
+			$update   = new JUpdate;
+			$instance = JTable::getInstance('update');
+			$instance->load($uid);
+			$update->loadFromXml($instance->detailsurl);
+
```
Comment (Quy) [inline administrator/components/com_installer/models/update.php:638]: Remove tabs

File: administrator/components/com_installer/models/update.php [hunk 22]
```diff
@@ -616,4 +616,144 @@ protected function preparePreUpdate($update, $table)
...
+			{
+				$app->enqueueMessage(\JText::sprintf('COM_INSTALLER_MANIFEST_INVALID_CERTIFICATE', $instance->name), 'warning');
+				continue;
+			}
+
+			$publickey = ParagonIE_Sodium_Compat::hex2bin($update->publickey->_data);
+			$signature = ParagonIE_Sodium_Compat::hex2bin($update->signature->_data);
+
+			$file = file_get_contents($instance->detailsurl);
+
```
Comment (Quy) [inline administrator/components/com_installer/models/update.php:671]: Remove tab

File: administrator/components/com_installer/models/update.php [hunk 23]
```diff
@@ -616,4 +616,144 @@ protected function preparePreUpdate($update, $table)
...
+			// Write the unsigned manfifest
+			$xml = $doc->savexml();
+ 			$a= file_put_contents(dirname(__DIR__) . '/unsignedmanifest.xml', $xml);
+
+			// Read the unsigned manifest
+			$manifest = file_get_contents(dirname(__DIR__) . '/unsignedmanifest.xml');
+
+			// Calculate hash digest
+			$digest   = hash("sha384", $manifest);
+
```
Comment (Quy) [inline administrator/components/com_installer/models/update.php:692]: Remove tab

**Comment by alikon** [comment]
travis&drone are happy now 
i hope that even human people will be more happy now :smile_cat:

**Pull request closed without merging**